### PR TITLE
New Module for SmartOS: nictagadm

### PR DESCRIPTION
**Pull request by Smithx10**
##### SUMMARY
This change adds the ability to create and destroy nic_tags on SmartOS.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
Module Name: nictagadm

##### ANSIBLE VERSION
```
(venv) bruce.smith@Bruces-MacBook-Pro /g/m/n/ansible ❯❯❯ ansible --version                                                                                              devel ⬇ ◼
ansible 2.5.0 (devel 273a3d1d51) last updated 2018/02/08 17:32:54 (GMT +700)
  config file = None
  configured module search path = [u'/Users/bruce.smith/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /git/me/nictagadm/ansible/lib/ansible
  executable location = /git/me/nictagadm/ansible/bin/ansible
  python version = 2.7.14 (default, Jan  6 2018, 12:16:16) [GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.39.2)]
```

### CODE REVIEW

File: lib/ansible/modules/cloud/smartos/nictagadm.py [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+version_added: "2.5"
+author: Bruce Smith (@SmithX10)
+options:
+    name:
+        description:
+            - NIC tag name.
+        required: true
+    mac:
+        description:
+            - MAC Address to attach the NIC tag to when not using an etherstub. mac and etherstub are mutually exclusive.
```
Comment (jasperla) [inline lib/ansible/modules/cloud/smartos/nictagadm.py]: Referring to other options would be better formatted as I(mac) and I(etherstub) to be consistent with some of the other SmartOS modules.

File: lib/ansible/modules/cloud/smartos/nictagadm.py [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+        required: false
+        default: None
+    etherstub:
+        description:
+            - Specifies that the NIC tag will be attached to a created etherstub. etherstub is mutually exclusive with both mtu, and mac.
```
Comment (jasperla) [inline lib/ansible/modules/cloud/smartos/nictagadm.py]: here too.

File: lib/ansible/modules/cloud/smartos/nictagadm.py [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+        required: false
+        default: false
+    mtu:
+        description:
+            - MTU size of the NIC tag. mtu and etherstub are mutually exclusive.
```
Comment (jasperla) [inline lib/ansible/modules/cloud/smartos/nictagadm.py]: and here.

File: lib/ansible/modules/cloud/smartos/nictagadm.py [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+        required: false
+        default: None
+    force:
+        description:
+            - When State.Absent is set this switch will use the -f parameter and delete the NIC tag regardless of existing VMs
```
Comment (jasperla) [inline lib/ansible/modules/cloud/smartos/nictagadm.py]: Would be better formatted as "When I(state) is absent this switch will use the C(-f) [...]" and there's a missing full stop.

File: lib/ansible/modules/cloud/smartos/nictagadm.py [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+    state:
+        description:
+            - Create or delete a SmartOS NIC tag.
+        required: false
+        default: "present"
+        choices: [ "present", "absent" ]
+'''
+
+EXAMPLES = '''
+- name: Create 'strage0' on '00:1b:21:a3:f5:4d'
```
Comment (jasperla) [inline lib/ansible/modules/cloud/smartos/nictagadm.py]: typo, should be 'storage0'

File: lib/ansible/modules/cloud/smartos/nictagadm.py [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+  nictagadm: name=storage0 mac=00:1b:21:a3:f5:4d mtu=9000 state=present
+
+- name: Remove 'storage0' NIC tag
+  nictagadm: name=storage0 state=absent force=false
```
Comment (jasperla) [inline lib/ansible/modules/cloud/smartos/nictagadm.py]: `force=false` is the default, so no need to explicitly add it, right?

File: lib/ansible/modules/cloud/smartos/nictagadm.py [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+    type: string
+    sample: present
+'''
+
+
+from ansible.module_utils.basic import AnsibleModule
+import re
+
+
+class NICTAG(object):
```
Comment (jasperla) [inline lib/ansible/modules/cloud/smartos/nictagadm.py]: This is a personal preference, but I'd rather not have a class name in all caps.

File: lib/ansible/modules/cloud/smartos/nictagadm.py [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+    result['state'] = nictag.state
+
+    if not nictag.is_valid_mac():
+        module.fail_json(msg='Invalid MAC Address Value',
+                         name=nictag.name,
+                         mac=nictag.mac,
+                         etherstub=nictag.etherstub,
+                         mtu=nictag.mtu,
+                         force=nictag.force,
+                         state=nictag.state)
```
Comment (jasperla) [inline lib/ansible/modules/cloud/smartos/nictagadm.py]: Isn't this a bit too much context to report merely the MAC access is invalid? I'd include the `name`, `mac` and `etherstub` and leave out the others.

File: lib/ansible/modules/cloud/smartos/nictagadm.py [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+
+    def nictag_exists(self):
+        cmd = [self.module.get_bin_path('nictagadm', True)]
+
+        cmd.append('exists')
+        cmd.append(self.name)
+
+        (rc, dummy, dummy) = self.module.run_command(cmd)
+
+        if rc == 0:
```
Comment (jasperla) [inline lib/ansible/modules/cloud/smartos/nictagadm.py]: Could be shortened to `return rc == 0` if you're not doing any further processing other than returning.

**Review by jasperla — CHANGES_REQUESTED** [review]
I've not tested this module yet, once you could fix up the comments above I'll try it out. Thanks for adding more SmartOS support!

**Comment by Smithx10** [comment]
@jasperla I made the suggested changes, please critique.  On a side note, do you think we need any other SmartOS particular modules?

**Review by jasperla — APPROVED** [review]
Looks good to me.

**Comment by jasperla** [comment]
@Smithx10 I don't particularly miss any other modules right now. That said, don't let that stop you from extending SmartOS support.

**Comment by jasperla** [comment]
shipit

File: lib/ansible/modules/cloud/smartos/nictagadm.py [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = '''
+---
+module: nictagadm
+short_description: Manage nic tags on SmartOS systems.
+description:
+  - Create of delete nic tags on SmartOS systems.
```
Comment (jasperla) [inline lib/ansible/modules/cloud/smartos/nictagadm.py]: s/of/or/

**Comment by maxamillion** [comment]
@Smithx10 any desire to support `check_mode` ?

**Comment by Smithx10** [comment]
@maxamillion,

I've been a little tied up lately, but I'll see what's involved in implementing check_mode.

Thank You,
Bruce Smith

**Comment by maxamillion** [comment]
@Smithx10 awesome!

**Review by dagwieers — APPROVED** [review]
LGTM, with a few cosmetic/doc changes

**Comment by dagwieers** [comment]
rebuild_merge

**Comment by dagwieers** [comment]
rebuild_merge

**Pull request merged**